pkg/cachemanager/cachemanager.go

```diff
@@ -0,0 +1,370 @@
+	}
+
+	// bail because it means we should not be
+	// syncing this gvk's objects as it is namespace excluded.
+	if isNamespaceExcluded {
+		c.tracker.ForData(instance.GroupVersionKind()).CancelExpect(instance)
+		return nil
+	}
+
+	syncKey := syncutil.GetKeyForSyncMetrics(instance.GetNamespace(), instance.GetName())
+	if c.watchedSet.Contains(gvk) {
+		_, err = c.opa.AddData(ctx, instance)
+		if err != nil {
+			c.syncMetricsCache.AddObject(
+				syncKey,
+				syncutil.Tags{
+					Kind:   instance.GetKind(),
+					Status: metrics.ErrorStatus,
+				},
+			)
+
+			return err
+		}
+	}
+
+	c.tracker.ForData(instance.GroupVersionKind()).Observe(instance)
+
+	c.syncMetricsCache.AddObject(syncKey, syncutil.Tags{
+		Kind:   instance.GetKind(),
+		Status: metrics.ActiveStatus,
+	})
+	c.syncMetricsCache.AddKind(instance.GetKind())
+
+	return err
+}
+
+func (c *CacheManager) RemoveObject(ctx context.Context, instance *unstructured.Unstructured) error {
+	gvk := instance.GroupVersionKind()
+
+	if c.watchedSet.Contains(gvk) {
+		if _, err := c.opa.RemoveData(ctx, instance); err != nil {
+			return err
+		}
+	}
+
+	// only delete from metrics map if the data removal was succcesful
+	c.syncMetricsCache.DeleteObject(syncutil.GetKeyForSyncMetrics(instance.GetNamespace(), instance.GetName()))
+	c.tracker.ForData(instance.GroupVersionKind()).CancelExpect(instance)
+
+	return nil
+}
+
+func (c *CacheManager) wipeData(ctx context.Context) error {
+	if _, err := c.opa.RemoveData(ctx, target.WipeData()); err != nil {
+		return err
+	}
+
+	// reset sync cache before sending the metric
+	c.syncMetricsCache.ResetCache()
+	c.syncMetricsCache.ReportSync()
+
+	return nil
+}
+
+func (c *CacheManager) ReportSyncMetrics() {
+	c.syncMetricsCache.ReportSync()
+}
+
+func (c *CacheManager) syncGVK(ctx context.Context, gvk schema.GroupVersionKind) error {
+	u := &unstructured.UnstructuredList{}
+	u.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   gvk.Group,
+		Version: gvk.Version,
+		Kind:    gvk.Kind + "List",
+	})
+
+	var err error
+	c.mu.Lock()
+	if !c.watchedSet.Contains(gvk) {
+		// we are not actually watching this gvk anymore
+		// so don't list instances for it.
+		err = nil
+	} else {
+		err = c.reader.List(ctx, u)
+	}
+	c.mu.Unlock()
+
+	if err != nil {
+		return fmt.Errorf("replaying data for %+v: %w", gvk, err)
+	}
+
+	for i := range u.Items {
+		if err := c.AddObject(ctx, &u.Items[i]); err != nil {
+			return fmt.Errorf("adding data for %+v: %w", gvk, err)
+		}
+	}
+
+	return nil
+}
+
+func (c *CacheManager) manageCache(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			close(c.stopChan)
+			return
+		case <-c.backgroundManagementTicker.C:
+			c.mu.Lock()
+			c.wipeCacheIfNeeded(ctx)
+
+			// spin up new goroutines to relist gvks as there has been a wipe
+			if c.needToList {
+				// stop any goroutines that were relisting before
+				// as we may no longer be interested in those gvks
+				c.stopChan <- true
```

Comment:
standard is to just close the channel

Reply:
since we close the channel above on `case <- ctx.Done()`, I'm thinking about how we can safely check that the channel has been closed on `ctx.Done()` for the `manageCache` func. 

That's why I opted to use a buffered channel w a value sent to tell the goroutine to stop